### PR TITLE
feat: Add `Option<T>` to noir stdlib

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Relates to #988, although this PR does not add `enum`s in general, nor does it add `match` to Noir. Only the `Option` type is provided along with some helper methods.

## Documentation

- [x] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [x] I will request for and support Dev Rel's help in documenting this PR.
    - We'll need to document the Option type along with each method on it. Each method I've added should match exactly the method of the same name on https://doc.rust-lang.org/std/option/enum.Option.html.
    - `Option`'s fields (`value` and `_is_some`) do not need to be documented, they can be considered private.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
I added a test for `option` but unfortunately it is causing a panic related to DecisionTree handling in the current SSA IR.

It works when running with `--experimental-ssa` so I think this PR is blocked until the ssa refactor is finished, assuming the decision tree bug is not fixed in the meantime.

**Comment by jfecher** [comment]
Ok, after some testing I've confirmed the panic happens when any of the methods accepting other functions as arguments are used. That is: `map`, `map_or`, `map_or_else`, `unwrap_or_else`, `and_then`, and `or_else`

**Comment by iAmMichaelConnor** [comment]
> so I think this PR is blocked until the ssa refactor is finished

Is this PR still blocked?
I was about to write a (less good) version of Option<T> in the Noir Contracts library of Aztec3, then stumbled upon this PR.

**Comment by jfecher** [comment]
This PR is no longer blocked since the old SSA code was removed. It is ready for review/merge now

**Comment by iAmMichaelConnor** [comment]
FYI I ended up taking a copy of this (thank you) and putting it in aztec-packages (temporarily). I had to make some edits to make it work over user-defined structs `T` - edits which make the interface deviate slightly from this one (mainly because `Option::none()` had to take a parameter for what "none" means for type `T`). I also added some comments, mostly copy-pasting the rust docs, because I wasn't sure what some of these methods were for (because I'm not familiar enough with Rust).

**Comment by jfecher** [comment]
@iAmMichaelConnor what were the issues that arose from using `std::unsafe::zeroed()` for the None case?

File: crates/nargo_cli/tests/test_data_ssa_refactor/option/Nargo.toml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[package]
```
Comment (vezenovm) [inline crates/nargo_cli/tests/test_data_ssa_refactor/option/Nargo.toml:1]: This test should be moved to `test_data` now right? I believe you will also have to add a `name` field.

**Review by vezenovm — COMMENTED** [review]
This PR looks to me for the most part. I mainly request that we add some comments on what each of the functions are supposed to do (I find that `and` on Options can be confusing if you are not familiar with Rust).

**Comment by vezenovm** [comment]
> I had to make some edits to make it work over user-defined structs `T`

Seeing Mike's comments now it would be good to check why `std::unsafe::zeroed()` may be causing problems

**Comment by iAmMichaelConnor** [comment]
> @iAmMichaelConnor what were the issues that arose from using std::unsafe::zeroed() for the None case?

I think I just got scared about using `std::unsafe::zeroed()` for struct types, because the docs (https://noir-lang.org/standard_library/zeroed/) warn against doing so (and because it's called "unsafe"). So I opted to encourage the user to pass as an argument an instance of what "none" means for their struct.

File: noir_stdlib/src/option.nr [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    fn map_or_else<U>(self, default: fn() -> U, f: fn(T) -> U) -> U {
+        if self._is_some {
+            f(self.value)
+        } else {
+            default()
+        }
+    }
+
+    fn and(self, other: Self) -> Self {
```
Comment (vezenovm) [inline noir_stdlib/src/option.nr:68]: Could you add some comments to these methods?

**Review by vezenovm — APPROVED** [review]

**Comment by iAmMichaelConnor** [comment]
Ah, ok, I misunderstood. In that case, I'll start using this stdlib Option type, instead of the modified version.

Re improving the docs wording. I'd maybe break the two examples (bounded vec and enumerations) into separate paragraphs, and perhaps add some code snippets to illustrate the points more clearly - i.e. to show how `zeroed` is being used and why it won't ever be accessed. Basically, I need my hand to be held a bit more, because these sentences assume some knowledge and experience that my brain doesn't have.

I know this PR has gone through, but might I suggest the `value` property of the `Option` struct is renamed to `_value`, if it should only ever be 'got' via an `unwrap` method? I actually encountered confusion when I was debugging the wrapping of some types in `Option` last week, because one of my types `T` also had a `value` method which I was trying to access.

Perhaps also could we add an `unwrap_unsafe()` which doesn't `assert(self._is_some())`? I found myself following a pattern of checking 
```
if my_option.is_some() {
    let my_thing = my_option.unwrap_unsafe();
    // do some other stuff
}
```

**Pull request merged**